[2/n] [c10d] Introduce the 'multiTenant' constructor parameter in TCPStore

cbalioglu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58331 [4/n] [c10d] Introduce the multi-tenancy feature in TCPStore
* #58330 [3/n] [c10d] Revise the implementation of TCPStore
* **#58329 [2/n] [c10d] Introduce the 'multiTenant' constructor parameter in TCPStore**
* #58328 [1/n] [c10d] Introduce a new TCPStore constructor

This PR is part of a stack that addresses the GitHub issue #41614; it introduces:

- A new `multiTenant` constructor option for the `TCPStore` class indicating whether multiple store instances can be initialized with the same host:port pair.

- Updates to the C10d distributed (elastic) rendezvous and the `init_process_group` method to leverage the new `multiTenant` feature.

Note that the multi-tenancy feature itself is implemented in the fourth PR of this stack. In this PR passing `true` to `multiTenant` results only with a warning output.

Differential Revision: [D28424978](https://our.internmc.facebook.com/intern/diff/D28424978/)

cbalioglu committed:
[2/n] [c10d] Introduce the 'multiTenant' constructor parameter in TCPStore

This PR is part of a stack that addresses the GitHub issue #41614; it introduces:

- A new `multiTenant` constructor option for the `TCPStore` class indicating whether multiple store instances can be initialized with the same host:port pair.

- Updates to the C10d distributed (elastic) rendezvous and the `init_process_group` method to leverage the new `multiTenant` feature.

Note that the multi-tenancy feature itself is implemented in the fourth PR of this stack. In this PR passing `true` to `multiTenant` results only with a warning output.

Differential Revision: [D28424978](https://our.internmc.facebook.com/intern/diff/D28424978/)

[ghstack-poisoned]

cbalioglu committed:
Update on "[2/n] [c10d] Introduce the 'multiTenant' constructor parameter in TCPStore"


This PR is part of a stack that addresses the GitHub issue #41614; it introduces:

- A new `multiTenant` constructor option for the `TCPStore` class indicating whether multiple store instances can be initialized with the same host:port pair.

- Updates to the C10d distributed (elastic) rendezvous and the `init_process_group` method to leverage the new `multiTenant` feature.

Note that the multi-tenancy feature itself is implemented in the fourth PR of this stack. In this PR passing `true` to `multiTenant` results only with a warning output.

Differential Revision: [D28424978](https://our.internmc.facebook.com/intern/diff/D28424978/)

[ghstack-poisoned]

rohan-varma commented on torch/distributed/distributed_c10d.py:
```diff
@@ -520,6 +520,10 @@ def init_process_group(backend,
 
+            # Use a PrefixStore to avoid accidental overrides of keys used by
+            # different systems (e.g. RPC) in case the store is multi-tenant.
+            store = PrefixStore("default_pg", store)
```
Since multitenancy is not introduced in this PR, does it make more sense for this change to go in the 4th PR of the stack? Although it seems fine here too as long as it doesn't create any issues.

cbalioglu replied:
See my comment below.

rohan-varma commented on torch/distributed/rendezvous.py:
```diff
@@ -139,7 +139,9 @@ def _error(msg):
     assert result.hostname is not None
-    store = TCPStore(result.hostname, result.port, world_size, start_daemon, timeout)
+    store = TCPStore(  # type: ignore[call-arg]
+        result.hostname, result.port, world_size, start_daemon, timeout, multi_tenant=True
```
Same question - does it make more sense to flip it to multitenant in the PR that enables it? Although shouldn't really make a difference if you're planning to land the whole stack at around the same time.

rohan-varma reviewed (approved):
Looks great, thanks!

cbalioglu committed:
Update on "[2/n] [c10d] Introduce the 'multiTenant' constructor parameter in TCPStore"


This PR is part of a stack that addresses the GitHub issue #41614; it introduces:

- A new `multiTenant` constructor option for the `TCPStore` class indicating whether multiple store instances can be initialized with the same host:port pair.

- Updates to the C10d distributed (elastic) rendezvous and the `init_process_group` method to leverage the new `multiTenant` feature.

Note that the multi-tenancy feature itself is implemented in the fourth PR of this stack. In this PR passing `true` to `multiTenant` results only with a warning output.

Differential Revision: [D28424978](https://our.internmc.facebook.com/intern/diff/D28424978/)

[ghstack-poisoned]

cbalioglu committed:
Update on "[2/n] [c10d] Introduce the 'multiTenant' constructor parameter in TCPStore"


This PR is part of a stack that addresses the GitHub issue #41614; it introduces:

- A new `multiTenant` constructor option for the `TCPStore` class indicating whether multiple store instances can be initialized with the same host:port pair.

- Updates to the C10d distributed (elastic) rendezvous and the `init_process_group` method to leverage the new `multiTenant` feature.

Note that the multi-tenancy feature itself is implemented in the fourth PR of this stack. In this PR passing `true` to `multiTenant` results only with a warning output.

Differential Revision: [D28424978](https://our.internmc.facebook.com/intern/diff/D28424978/)

[ghstack-poisoned]

cbalioglu committed:
Update on "[2/n] [c10d] Introduce the 'multiTenant' constructor parameter in TCPStore"


This PR is part of a stack that addresses the GitHub issue #41614; it introduces:

- A new `multiTenant` constructor option for the `TCPStore` class indicating whether multiple store instances can be initialized with the same host:port pair.

- Updates to the C10d distributed (elastic) rendezvous and the `init_process_group` method to leverage the new `multiTenant` feature.

Note that the multi-tenancy feature itself is implemented in the fourth PR of this stack. In this PR passing `true` to `multiTenant` results only with a warning output.

Differential Revision: [D28424978](https://our.internmc.facebook.com/intern/diff/D28424978/)

[ghstack-poisoned]

cbalioglu committed:
Update on "[2/n] [c10d] Introduce the 'multiTenant' constructor parameter in TCPStore"


This PR is part of a stack that addresses the GitHub issue #41614; it introduces:

- A new `multiTenant` constructor option for the `TCPStore` class indicating whether multiple store instances can be initialized with the same host:port pair.

- Updates to the C10d distributed (elastic) rendezvous and the `init_process_group` method to leverage the new `multiTenant` feature.

Note that the multi-tenancy feature itself is implemented in the fourth PR of this stack. In this PR passing `true` to `multiTenant` results only with a warning output.

Differential Revision: [D28424978](https://our.internmc.facebook.com/intern/diff/D28424978/)

[ghstack-poisoned]

cbalioglu committed:
Update on "[2/n] [c10d] Introduce the 'multiTenant' constructor parameter in TCPStore"


This PR is part of a stack that addresses the GitHub issue #41614; it introduces:

- A new `multiTenant` constructor option for the `TCPStore` class indicating whether multiple store instances can be initialized with the same host:port pair.

- Updates to the C10d distributed (elastic) rendezvous and the `init_process_group` method to leverage the new `multiTenant` feature.

Note that the multi-tenancy feature itself is implemented in the fourth PR of this stack. In this PR passing `true` to `multiTenant` results only with a warning output.

Differential Revision: [D28424978](https://our.internmc.facebook.com/intern/diff/D28424978/)

[ghstack-poisoned]
